### ISSUE TITLE
Swap in new harddrives

### ISSUE THREAD

**Issue by mlissner** [freelawproject/courtlistener#693]
We've got the old drive mounted at `/sata` and the new drive mounted at `/sata8`. 

There are two important chunks of data to migrate:

1. The solr index for RECAP (about 400GB)
2. Gazillions of files.

The process here will be:

# Solr

1. Create a new empty index on the new drive ("recap-new")
1. Reindex the RECAP content to the new index

        cl && sudo -H -u www-data ../.virtualenvs/courtlistener/bin/python ./manage.py cl_update_index --solr-url http://127.0.0.1:8983/solr/recap-new --update --everything --optimize --type recap-dockets --traceback ; bush "Finished recap indexing"

1. Create a symlink from the old solr location to the new.
1. Re-index anything that changed during step 2.

# Files

1. Rsync all the files from old to new one directory at a time.

    The command for this is something like:

        rsync -azh --info-progress=2 /sata/ /sata8

    Though we'll want to exclude the solr index when we do this next time.

1. Set up symlinks from old to new for every directory.

# Finally

1. Put site in maintenance mode.
1. Stop celery and solr.
1. Take down `/sata` and `/sata8`
1. Bring up the new drive as `/sata`
1. Start celery and solr
1. Take site out of maintenance mode.
1. Delete `/usr/local/solr/example/solr/recap-new` (it points to `/sata8`, which no longer exists).

**Comment by mlissner** [comment]
OK, the problem above is now fixed thanks to Brian, so I started the process of re-indexing all of recap solr. It's cruising along right now at some incredible pace. (This is wildly fast because of the new Celery Throttle tool, which removed the majority of polling from our Celery tasks, and also because we have *both* servers chugging on this.)

I'll get a push notification when this finishes, at which point I'll proceed to the next step.

**Comment by mlissner** [comment]
We're now 6% done and using about 100GB of space for the index. This seems off, but I'm hopeful this is just a reflection of the bigger dockets taking up more space and being indexed first. The old index is around 300GB.

**Comment by mlissner** [comment]
The reindex is complete at 314GB of index files containing 16.8M documents, compared to 326GB for the current index containing 17.2M docs. I'm not sure why the reindex has fewer documents than the original, but the size going down makes sense according to #660. 

Tomorrow I'll do the symlink work and after that we should be ready to swap /sata8 an /sata.

Here's the output from the reindex:

```text
↪  cl && sudo -H -u www-data ../.virtualenvs/courtlistener/bin/python ./manage.py cl_update_index --solr-url http://127.0.0.1:8983/solr/recap-new --update --everything --optimize --type recap-dockets --traceback ; bush "Finished recap indexing"
[sudo] password for mlissner:
Running in update mode...
Adding or updating all items...
Processed 1795594/1795594 (100%)
Completed in 58511.5 seconds.
Optimizing the index...
done.
Completed in 48416.5 seconds.
```

Total time was about 29h, so we averaged 160 items/s, which is about 580k/hour. I can live with that.

**Comment by mlissner** [comment]
Performance of the new drive looks great as always:

```text
↪ sudo hdparm -Tt /dev/mapper/vg7-lvsata 

/dev/mapper/vg7-lvsata:
 Timing cached reads:   16136 MB in  2.00 seconds = 8075.18 MB/sec
 Timing buffered disk reads: 670 MB in  3.00 seconds = 223.20 MB/sec
10:57:36::mlissner@new-courtlistener::/sata 
```

Compared to the old drive, it's a nice improvement.

```text
↪ sudo hdparm -Tt /dev/mapper/vg5-lv0 

/dev/mapper/vg5-lv0:
 Timing cached reads:   14744 MB in  2.00 seconds = 7379.22 MB/sec
 Timing buffered disk reads: 456 MB in  3.01 seconds = 151.60 MB/sec
```

**Comment by mlissner** [comment]
This command will do the solr swap:

    sudo service solr stop && sudo mv /sata/solr-data /sata/solr-data.orig && sudo ln -s /sata8/solr-data/ /sata/solr-data && sudo service solr start

And if there are issues, we can undo it with:

    sudo service solr stop && sudo rm /sata/solr-data && sudo mv /sata/solr-data.orig /sata/solr-data && sudo service solr start

**Comment by mlissner** [comment]
Well...that didn't work. I don't know what happened, but the new index is now only 14GB:

```text
11:10:33::mlissner@new-courtlistener::/sata8/solr-data 
↪ du -h
14G	./tlog
96K	./index
14G	.
```

It could be that when I started solr it deleted any parts of the index that it didn't recognize? I've restored things to the old drive and will be asking on the Solr mailing list.

**Comment by mlissner** [comment]
I'm going to try this again, but for now I'm going to just copy the directory until I hear back from the solr list:

     rsync -azh --delete --info=progress2 /sata/solr-data/ /sata8/solr-data

When that's done, we'll have an out of sync solr directory, which we can fix with:

1. Stop solr cleanly
2. Sync again:

         rsync -azh --checksum --delete --info=progress2 /sata/solr-data/ /sata8/solr-data

    Note that `--checksum` is recommended by the postgres folks because otherwise the modtime used by rsync is only granular to 1s intervals...not quite enough.

1. Set up the symlink again:

        sudo mv /sata/solr-data /sata/solr-data.orig && sudo ln -s /sata8/solr-data/ /sata/solr-data

3. Start solr

**Comment by mlissner** [comment]
OK, rsync command is complete for the index.

Running the following:

    sudo service solr stop && sleep 10 && sudo rsync -azh --delete --info=progress2 /sata/solr-data/ /sata8/solr-data && sudo mv /sata/solr-data /sata/solr-data.orig && sudo ln -s /sata8/solr-data/ /sata/solr-data && sudo service solr start

Solr is up and running. Onwards to the various directories.

**Comment by mlissner** [comment]
Starting with a very big re-sync of everything:

    sudo rsync -azh --delete --info=progress2 /sata/ /sata8 --exclude /sata/solr-data*

When that's done, I'll do the following for each directory:

    sudo mv $DIR $DIR.org && sudo ln -s /sata8/$DIR/ /sata/$DIR && sudo rsync -azh --info=progress2 /sata/$DIR /sata8/$DIR

That'll move the item, then resync any new items that came in during the last sync, without deleting any new items that come into the new location in the process of the sync.

**Comment by mlissner** [comment]
Well, the rsync command above was successful in deleting the solr core on the new drive, and I had to first serve PACER results directly from our backup, then do a reindex. This kept things live by serving from a stale index, but this process is not going well. 

That said, at this point, the solr index is in pretty good shape, except that it seems to be missing a bit of content, as described in issue #696. 

Gotta wrap this issue up. It seems to be cursed.

**Comment by mlissner** [comment]
OK, just ran this against every directory in /sata: 

    export v='docx' && cd /sata && sudo mv $v $v.orig && sudo ln -s /sata8/$v $v && sudo chown -h www-data:www-data /sata/$v && sudo rsync -azh --delete --info=progress2 /sata/$v/ /sata8/$v/

In english:

 - backup the dir to .orig
 - set up a symbolic link to /sata8
 - get the ownership right on the symbolic link
 - sync any files that need to be synced.

The `recap` directory is going now. The rest are done.

**Comment by mlissner** [comment]
OK, things are mostly OK today. We were getting errors from the celery task server because setting up a symlink broke the NFS share. I got that up and running again, but now it's linking directly to /sata8 instead of /sata. 

So....I think we're in the home stretch. Here's what remains:

 - [ ] Stop Celery on new and old servers by running `sudo service celeryd stop` on each.
 - [ ] Stop Solr with `sudo service solr stop` (THINGS ARE NOW DOWN)
 - [ ] Remap /sata8 to /sata
 - [ ] Tweak solr configuration at `/usr/local/solr-4.10.4/example/solr/recap-new/core.properties` to have the right dataDir value.
 - [ ] Move the solr configuration at `/usr/local/solr-4.10.4/example/solr/recap` out of the solr directory (this will remove the core while backing up the directory).
 - [ ] Start solr with `sudo service solr start` (watch the logs -- THINGS ARE NOW UP)
 - [ ] Map the NFS share on the celery task server back to /sata with:

    - Client: Unmount the old NFS: `sudo umount 192.168.0.122:/sata8`
    - Client: Edit fstab to comment out the new line and replace it with the old one (that's currently commented out).
    - Client: Remount the /sata location with: `sudo mount 192.168.0.122:/sata /sata`
    - Host: Edit /etc/exports to remove line about /sata8
    - Host: Run `sudo exportfs -a`
    - Host: Restart NFS: `sudo service nfs-kenel-server restart`

 - [ ] Restart celery on all servers with `sudo service celeryd start`

**Comment by mlissner** [comment]
I just did some work (saved locally in drive-info.md) to figure out how the disks all work with RAID, LVM, and mount points. I'm going to wrap up this issue today, if I can.

The problems I identified (as listed in drive-info.md) are:

1. md6 should have four disks, but instead lists two in mdstat. The two disks that are listed, sdl and sdi, aren't listed in lshw (do not exist?). md6 is also not mounted to any vg?

2. md4 has a failing disk, and it's volume group (vg4) isn't mounted. 

3. /sources is listed in /etc/fstab, but isn't mounted

4. /sata is about to go away.

I'd attach drive-info.md, but Github doesn't support md files for some bloody reason.

**Comment by mlissner** [comment]
OK, this last step seems to have gone pretty well. The only issue I ran into was that for some reason I couldn't unmount the old /sata mountpoint. The issue appears to be related to it being part of an NFS mount. 

Part of the process above was to take down the NFS mount and move it to the new location, but that was planned after swapping /sata and /sata8. The order of these steps needed to be reversed for this to work though, since I couldn't unmount without first taking down the NFS mount. So I skipped ahead a bit, took down the NFS, then was able to do the umount of sata if I used the lazy mode. 

With regards to taking down the NFS mount, I followed the instructions here, but they didn't work:

https://unix.stackexchange.com/questions/15024/umount-device-is-busy-why

Part of those instructions was to take down the NFS mount, so that was done. The only other option I could think of was to force unmount it, so I did a "lazy" umount with:

    sudo umount -l /sata

That worked, and I was able to proceed. With that done, I put the NFS mount back up, started celery again, and monitored logs. Things seem fine from here.

CLOSING.
